### PR TITLE
nixos/bind: allow zone declaration in nix expression

### PR DESCRIPTION
**Pull request by paumr**
###### Motivation for this change
This PR will allow the declaration of zone-files with a nix expression.

Minimal example:
``` nix
        {
          name = "minimal";
          file = {
            records = [
              { name = "ns"; value = "192.168.1.1"; }
            ];
          };
        }
```

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/bind.nix [hunk 1]
```diff
@@ -33,31 +50,176 @@ let
...
-          ({ name, file, master ? true, slaves ? [], masters ? [], extraConfig ? "" }:
+      (
+        { name, file, master, slaves, masters, extraConfig }:
+        # TODO: there has to be a better way:
+        # generate zone-file if file is an attrset, else use file
+        # checks if the attrs isn't a derivation incase something like pkgs.writeText was assigned to the file.
+          ''
+            zone "${name}" {
+              type ${if master then "master" else "slave"};
+              file "${if (builtins.isAttrs file) && !(isDerivation file) then zoneFile { inherit name; opts = file; } else file}";
```
Comment (paumr) [inline nixos/modules/services/networking/bind.nix:61]: Is there a better way to solve this?

`file` was previously untyped, thus `str`, `path`, or `derivation` (which result in a path) was accepted (maybe even more types?)
Now `file` should allow the previous types OR the new submodule, what's the best way to check which one it is?

File: nixos/modules/services/networking/bind.nix [hunk 2]
```diff
@@ -103,15 +265,15 @@ in
       };
 
       listenOn = mkOption {
-        default = ["any"];
```
Comment (peti) [inline nixos/modules/services/networking/bind.nix:106]: Please don't mix cosmetic changes into the same PR with actual code changes as this makes the diffs uncessarily long and hard to review.

**Review by peti — CHANGES_REQUESTED** [review]
No reformatting noise, please.

File: nixos/modules/services/networking/bind.nix (hunk 2 above)
Comment (paumr) [reply nixos/modules/services/networking/bind.nix:106]: Sorry, i assumed separating them on a commit-level was enough. Should i close this PR in the meantime? (note that i initially set it to WIP)

Comment (paumr) [reply nixos/modules/services/networking/bind.nix:106]: fixed with #95050 , will rebase once #95050 was merged.

**Comment by vcunat** [comment]
No hard opinion, but I'd probably put tools for zone-file generation separately, as they could be used more universally (typically in other DNS servers).

:thinking: though perhaps I'm personally still unclear why
```nix
{ name = "ns"; value = "192.168.1.1"; }
```
seems easier than
```nix
"ns A 192.168.1.1"
```
(using interpolations if need be)

**Comment by paumr** [comment]
> No hard opinion, but I'd probably put tools for zone-file generation separately, as they could be used more universally (typically in other DNS servers).

Where would you put it?

> thinking though perhaps I'm personally still unclear why
> 
> ```nix
> { name = "ns"; value = "192.168.1.1"; }
> ```
> 
> seems easier than
> 
> ```nix
> "ns A 192.168.1.1"
> ```
> 
> (using interpolations if need be)

The idea was to abstract away the implementation details (aka. zone file).
This way:
* the user can rely on nix, rather than reading through the zone-file specs before setting up a DNS server
* the user input can be type-checked

Reasoning on why not `nix A 192.168.1.1`
* As a user i wouldn't be sure if this implies a leading whitespace or not.
* As a user i'd have to check the file specs to see if it's possible to mix Tabs/Spaces, and check the module source to see which one is used

Also: This format works quite nicely with more complex setups, e.g.:
``` nix
  records = map # add default type
    (x: x // { type = "A"; }) [
    { name = "toaster"; value = "10.0.3.1"; }
    { name = "localnews"; value = ip; }
    { name = "router"; value = ip; }
    { name = "ns"; value = ip; }
  ] ++ # reserved devices (aka poor man's zeroconf)
  map (x: { name = x.hostName; target = x.ipAddress; }) machines;
```

**Comment by vcunat** [comment]
> the user input can be type-checked

I'd rather have the zone checked by a DNS server (during build time of the configs).

Overall it's perhaps a personal preference of not introducing yet another format unless there's some significant benefit, but I might be a little biased due to knowing the zonefile format (though I know lots of nix as well).

> Also: This format works quite nicely with more complex setups

I'm not sure whether it's such a big difference; I think nix is quite good with string manipulation, though perhaps those functions are a little more cumbersome:
```nix
lib.concatMapStringsSep "\n" (x: "${x.hostName} A ${x.ipAddress}")
```

**Comment by paumr** [comment]
> > the user input can be type-checked
> 
> I'd rather have the zone checked by a DNS server (during build time of the configs).

The DNS server still checks the format, this merely adds an additional check during compile-time (to prevent failure upon deployment).
 
> Overall it's perhaps a personal preference of not introducing yet another format unless there's some significant benefit, but I might be a little biased due to knowing the zonefile format (though I know lots of nix as well).
> > Also: This format works quite nicely with more complex setups
> 
> I'm not sure whether it's such a big difference; I think nix is quite good with string manipulation, though perhaps those functions are a little more cumbersome:
> 
> ```nix
> lib.concatMapStringsSep "\n" (x: "${x.hostName} A ${x.ipAddress}")
> ```

I guess you are right, it's a matter of preference, everyone can easily write his/her own zone-file generator on a per-project basis. 

To add a comparison of the two configuration methods (both would be still possible after the change):
``` nix
{
  name = ".";
  file =
    let
      enries = [
        { name = "ns1"; value = "192.168.1.1"; }
        { name = "webserver"; value = "192.168.1.1"; }
      ];
      primary_name = "ns1.example.org";
      mail = "admin.example.org";
    in
      pkgs.writeText "root.zone" ''
        $TTL 3600
        . IN SOA ${primary_name}. ${mail}. ( 1 8h 2h 4w 1d )
        . IN NS ${primary_name}.
        ${lib.concatMapStringsSep "\n" (x: "${x.name} A ${x.value}") entries}
      '';
}
```

``` nix
{
  name = "example.org";
  file = {
    serial = 1;
    name_primary = "ns1";
    records = [
      { name = "ns1"; value = "192.168.1.1"; }
      { name = "webserver"; value = "192.168.1.1"; }
    ];
  };
}
```
Note: this is pseudo code (aka i didn't run it)

**Comment by vcunat** [comment]
> > No hard opinion, but I'd probably put tools for zone-file generation separately, as they could be used more universally (typically in other DNS servers).
> 
> Where would you put it?

Probably somewhere inside the `lib` attribute (i.e. `lib/*.nix`), but I don't know if there would be consensus about it in the community.  It's true it seems way more specific than what I see in there now.

File: nixos/modules/services/networking/bind.nix (hunk 2 above)
Comment (paumr) [reply nixos/modules/services/networking/bind.nix]: #95050 was merged, i rebased on the current master, however due to this PR being stuck for a year the diff is now slightly messier than before (due to upstream changes)

**Pull request closed without merging**